src/NHibernate/Driver/OracleDataClientDriverBase.cs

```diff
@@ -0,0 +1,181 @@
+using System;
+using System.Collections.Generic;
+using System.Data;
+using System.Data.Common;
+using System.Reflection;
+using NHibernate.AdoNet;
+using NHibernate.Engine.Query;
+using NHibernate.SqlTypes;
+using NHibernate.Util;
+
+namespace NHibernate.Driver
+{
+	/// <summary>
+	/// A NHibernate driver base for using ODP.Net.
+	/// </summary>
+	/// <remarks>
+	/// Original code was contributed by <a href="http://sourceforge.net/users/jemcalgary/">James Mills</a>
+	/// on the NHibernate forums in this 
+	/// <a href="http://sourceforge.net/forum/message.php?msg_id=2952662">post</a>.
+	/// </remarks>
+	public abstract class OracleDataClientDriverBase : ReflectionBasedDriver, IEmbeddedBatcherFactoryProvider
+	{
+		private const string _commandClassName = "OracleCommand";
+
+		private const string _unmanagedProviderInvariantName = "Oracle.DataAccess.Client";
+		private const string _unmanagedDriverAssemblyName = "Oracle.DataAccess";
+		private const string _unmanagedClientNamespace = "Oracle.DataAccess.Client";
+
+		private const string _managedProviderInvariantName = "Oracle.ManagedDataAccess.Client";
+		private const string _managedDriverAssemblyName = "Oracle.ManagedDataAccess";
+		private const string _managedClientNamespace = "Oracle.ManagedDataAccess.Client";
+
+		private static readonly SqlType _guidSqlType = new SqlType(DbType.Binary, 16);
+		private readonly PropertyInfo _oracleCommandBindByName;
+		private readonly PropertyInfo _oracleDbType;
+		private readonly object _oracleDbTypeRefCursor;
+		private readonly object _oracleDbTypeXmlType;
+		private readonly object _oracleDbTypeBlob;
+		private readonly object _oracleDbTypeNVarchar2;
+		private readonly object _oracleDbTypeNChar;
+
+		/// <summary>
+		/// Default constructor.
+		/// </summary>
+		/// <param name="managed"><see langword="true"/> for loading the managed driver, <see langword="false"/> for loading the unmanaged driver.</param>
+		/// <exception cref="HibernateException">
+		/// Thrown when the <c>Oracle.DataAccess</c> assembly can not be loaded.
+		/// </exception>
+		protected OracleDataClientDriverBase(bool managed)
+			: this(
+				managed ? _managedProviderInvariantName : _unmanagedProviderInvariantName,
```

Comment:
Can we move these back to the constructor instead?

Reply:
I do not get it.

Would you like this to be removed and having manage and un-managed classes providing these values themselves? Or should the `ReflectionBasedDriver` be changed for allowing specifying these values without a constructor call (which forces to put that inside `: this(...)` or `: base(...)`), by example with some protected settable properties or a protected initialize method?
Or something else?

Reply:
What I meant is to receive these strings from the child constructor instead of a bool argument. Eg. move this constructor down the inheritance chain:
```
public OracleManagedDataClientDriver : base(_managedProviderInvariantName, ...)
```

Also, it seems that we need to have only 2 variable strings: "Oracle.ManagedDataAccess" and "Oracle.DataAccess", all other are derivatives

Reply:
Done, with only assembly name as a parameter. My previous idea was to be able to adapt to eventual weird naming changes done on Oracle side for some future version. But well, they may not do same than Sap with its Sybase drivers.